server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java

```diff
@@ -1632,12 +1632,83 @@ protected NumericSyntheticFieldLoader(String name, String simpleName) {
 
         @Override
         public Leaf leaf(LeafReader reader, int[] docIdsInLeaf) throws IOException {
-            SortedNumericDocValues leaf = dv(reader);
-            if (leaf == null) {
+            SortedNumericDocValues dv = dv(reader);
+            if (dv == null) {
                 return SourceLoader.SyntheticFieldLoader.NOTHING_LEAF;
             }
-            return new SourceLoader.SyntheticFieldLoader.Leaf() {
-                private boolean hasValue;
+            if (docIdsInLeaf.length > 1) {
+                /*
+                 * The singleton optimization is mostly about looking up all
+                 * values for the field at once. If there's just a single
+                 * document then it's just extra overhead.
+                 */
+                NumericDocValues single = DocValues.unwrapSingleton(dv);
+                if (single != null) {
+                    return singletonLeaf(single, docIdsInLeaf);
+                }
+            }
+            return new ImmediateLeaf(dv);
+        }
+
+        private class ImmediateLeaf implements Leaf {
+            private final SortedNumericDocValues dv;
+            private boolean hasValue;
+
+            ImmediateLeaf(SortedNumericDocValues dv) {
+                this.dv = dv;
+            }
+
+            @Override
+            public boolean empty() {
+                return false;
+            }
+
+            @Override
+            public boolean advanceToDoc(int docId) throws IOException {
+                return hasValue = dv.advanceExact(docId);
+            }
+
+            @Override
+            public void write(XContentBuilder b) throws IOException {
+                if (false == hasValue) {
+                    return;
+                }
+                if (dv.docValueCount() == 1) {
+                    b.field(simpleName);
+                    writeValue(b, dv.nextValue());
+                    return;
+                }
+                b.startArray(simpleName);
+                for (int i = 0; i < dv.docValueCount(); i++) {
+                    writeValue(b, dv.nextValue());
+                }
+                b.endArray();
+            }
+        }
+
+        /**
+         * Load all values for all docs up front. This should be much more
+         * disk and cpu-friendly than {@link ImmediateLeaf} because it resolves
+         * the values all at once, keeping the disk .
```

Comment:
keeping the disk ... in suspense?

Reply:
https://www.youtube.com/watch?v=wlwnbcxBuzI